compiler/rustc_expand/src/proc_macro.rs

```diff
@@ -133,6 +133,19 @@ impl MultiItemModifier for ProcMacroDerive {
             }
         }
 
+        // check if the entire stream was parsed
+        if parser.token != token::Eof {
+            ecx.struct_span_err(
+                parser.token.span,
+                &format!(
+                    "expected item, found {}",
+                    rustc_parse::parser::token_descr(&parser.token)
+                ),
+            )
+            .span_label(parser.token.span, "expected item")
+            .emit();
```

Comment:
I _think_ you could use [`parser.unexpected()`](https://searchfox.org/rust/rev/1c66d11a34047be1eb6c50703f8ba6689a15e716/compiler/rustc_parse/src/parser/mod.rs#477-484) here to get the parser to compose the error mentioning the unexpected token and what tokens would be expected, rather than generating an error manually. I think it'd look something roughly like:

```rust
if parser.token != token::Eof {
    parser.unexpected().unwrap_err().emit();
}
```

The other callers tend to be invoked on contexts returning a `PResult`, like https://searchfox.org/rust/rev/1c66d11a34047be1eb6c50703f8ba6689a15e716/compiler/rustc_builtin_macros/src/assert.rs#147, so I don't know if there's a better tool for this situation which doesn't require the `unwrap_err()`.

Reply:
Thanks!

Using that function, it produces:

```
error: expected one of `#`, `async`, `auto`, `const`, `crate`, `default`, `enum`, `extern`, `fn`, `impl`, `macro_rules`, `macro`, `mod`, `pub`, `static`, `struct`, `trait`, `type`, `unsafe`, `use`, or path, found `{`

        LL | #[derive(B)]
	   |          ^ expected one of 21 possible tokens
```

It's a bit verbose, but it looks accurate.

Reply:
Went with a different approach and split part of the existing `parse_mod` into a new `parse_items` so it could be re-used here. That way we don't have to manually produce the error here.